### DISCUSSION TITLE
Taxonomy Editor

### DISCUSSION THREAD

**Discussion by santhosh-tarento** [Sunbird-Knowlg/Community#168] [Feature Requests]
Definition:

What is taxonomy ?
     Taxonomy helps to clarify our thinking by classifying things neatly into categories and sub-categories based on their relationships. The 
      primary goal of taxonomy is to recognize, characterise, classify, and name based on their properties.

What is term ?
    Terms are items (entities or objects) with in the taxonomy, As below example shows education system    taxonomy, where CBSC, Hindi, 
     Grade1, English are terms belonging to different categories,

What is Association ?
    In the taxonomy system terms belonging to different category relate to each other, that relation is called association, each category term 
    might have multiple relations with terms belonging to different categories.

In the below example CBSC board have relate to Hindi medium,  like Hindi medium might have grde 1 and grade 2 relation.

Example:-   Education Taxonomy would look as below.

![image-20230405-170836](https://github.com/Sunbird-Knowlg/Community/assets/120555456/6d23ce21-9284-4f5b-bf1b-5c6627484012)

Problem Statement:  

In sun-bird system classification of terms and identify the relations (association) between terms belongs to different categories and show that in visual representation help user to creation and publishing of framework with categories.

Existing System

Current system of Taxonomy classifications are hierarchical representation of terms belongs to different category in a tree structure which might be complex when there are large system, It is difficult to show relation or association between terms belong to different category
creation and association of terms in sun bird is done through postman collection, 

Proposed System of Taxonomy

In proposed the taxonomy editor system under category we can create terms and associate/relate terms belonging to different categories in a visual representation as shown below this. 

![clip-board-1](https://github.com/Sunbird-Knowlg/Community/assets/120555456/14732b8c-8047-41b2-b7d9-200caf218171)

UI setup guide :- 

1. Install latest version of angular.

     npm i sb-taxonomy-editor

3.  Add below element in sunbird-Ed-portal
  
      <lib-taxonomy-view [environment]="environment" [taxonomyConfig]="taxonomyConfig"></lib-taxonomy-view>

3.  Save below  "environment"  and "taxonomyConfig" in local storage before initialize above library. (this need to change as input to a library).
                  
                 environment =  {
                         frameworkName: string,   
                         channelId: string,
                         authToken: string,
                         isApprovalRequired:  boolen  // set default to false,
                  }
     
                 taxonomyConfig = {
                            frameworkId: string,
                            config: [
                                {
                                    index:number,
                                    category:string,
                                    icon: 'string',
                                    color: 'string'
                                }
                            ]
                }

NPM Link:
   https://www.npmjs.com/package/sb-taxonomy-editor

**Comment by pallakartheekreddy** [comment]
@santhosh-tarento Thank you for sharing this thread. Could you kindly share additional documents, such as the technical design document and workflow, like to understand the list of APIs utilized by the editor and the sequence in which these APIs are called. 

@vrayulu @SanthoshVasabhaktula @coolbung @mathewjpallan @mohitga @kameshbhr @madhucr kindly request you to review the new RFC and share your inputs and feedback within a week.

**Comment by vrayulu** [comment]
@santhosh-tarento - This is great. Please share a video recording or UI screenshots of the taxonomy editor workflow to understand the full capabilities of the system.

I am assuming this is a standalone tool that can connect any running instance of Sunbird taxonomy service and enable UI based taxonomy management. Please confirm.

**Reply by thillaibirla** [reply]
Hi @vrayulu, yes its build as a stand alone tool(library) which can connect in any running instance of Sunbird Taxonomy service.  we will share the screenshots with you.

**Reply by thillaibirla** [reply]
@santhosh-tarento can you pls share a demo video covering the capabilities.

**Comment by santhosh-tarento** [comment]
@vrayulu @thillaibirla @pallakartheekreddy 
https://github.com/Sunbird-Knowlg/Community/assets/120555456/6b7a2e52-492f-4567-9f6b-6ae818c65861

pls find the video and document link with list of api's used

https://project-sunbird.atlassian.net/wiki/spaces/SBDES/pages/3302916100/Taxonomy+editor

**Comment by thillaibirla** [comment]
@pallakartheekreddy @vrayulu - can you pls review

**Reply by pallakartheekreddy** [reply]
@thillaibirla @santhosh-tarento Apologies for the delayed response. 

I have added my comments on the above wiki page, Please have a look and address them accordingly.

One more observation, that we are making a significant number of API calls to create a single term. Could we consider minimizing these calls? It seems that we are making a Framework publish call to obtain the hierarchy. Would it be possible to use a search API call as an alternative solution to address this issue?

@vrayulu @mohitga

**Reply by santhosh-tarento** [reply]
@pallakartheekreddy Regarding multiple api calls I have reduce multiple publish api calls, however newly created term need to associate with each parent category term, so these call's are need as there are not supporting api's update all the category term in single api.

**Comment by santhosh-tarento** [comment]
@pallakartheekreddy thanks for the feedback,  I will be looking into it today, I will comeback to you by Monday

**Comment by santhosh-tarento** [comment]
@pallakartheekreddy I have addressed comments. pls check

**Reply by pallakartheekreddy** [reply]
@santhosh-tarento Kindly fork the repository below, create a directory named `taxonomy-editor` integrate your code, and submit a pull request. Please add Readme.md file with installation steps.
https://github.com/Sunbird-Knowlg/knowledge-platform-tools

@mohitga

**Reply by santhosh-tarento** [reply]
@pallakartheekreddy https://github.com/deviigopinath/knowledge-platform-tools-fork/pull/1

**Reply by santhosh-tarento** [reply]
Pls find the PR

**Reply by pallakartheekreddy** [reply]
@santhosh-tarento Can you send PR to below repo. I guess you shared above your fork link.
https://github.com/Sunbird-Knowlg/knowledge-platform-tools

**Reply by santhosh-tarento** [reply]
@pallakartheekreddy pls add devi.gopinath@tarento.com and santhosh.kumar@tarento.com as collaborator to push
https://github.com/Sunbird-Knowlg/knowledge-platform-tools

**Reply by pallakartheekreddy** [reply]
@santhosh-tarento This is a public repo. Just ask them to fork and send PR.

**Reply by santhosh-tarento** [reply]
@pallakartheekreddy can you pls help with error in this PR https://github.com/Sunbird-Knowlg/knowledge-platform-tools/pull/2

**Reply by pallakartheekreddy** [reply]
@santhosh-tarento added review comments on above PR, please check and close.

**Reply by santhosh-tarento** [reply]
@pallakartheekreddy PR comments have been taken care, please let  me know if any changes are needed

**Reply by HarishGangula** [reply]
@santhosh-tarento  Can you check the latest PR review comments?

**Reply by santhosh-tarento** [reply]
@HarishGangula Team is looking into it, will update shortly

**Reply by santhosh-tarento** [reply]
@HarishGangula your comments have been addressed,  can you pls have look.

**Reply by HarishGangula** [reply]
@santhosh-tarento Still the code including the HTML template files contains commented code, can we address all those issues with basic sanity and then send it for review?
cc: @pallakartheekreddy @mohitga

**Reply by santhosh-tarento** [reply]
@HarishGangula, changes are done, please have look into it.

**Reply by thillaibirla** [reply]
@HarishGangula - whats the update on the PR?

**Reply by HarishGangula** [reply]
@thillaibirla  -  I have added the comments in the PR
Highlevel
1. Unit test cases to be written in basic flows which are critical
2. Unused and commented code need to be removed

**Reply by santhosh-tarento** [reply]
@HarishGangula  coverage is above 75%, Also Devi has addressed all your comments,  can you please add comment on where that basic flow is missing unit test case, I mean which flow you are are referring.

**Reply by HarishGangula** [reply]
@santhosh-tarento one example file is [this](https://github.com/Sunbird-Knowlg/knowledge-platform-tools/pull/2/files#diff-6904fdcfc78464b56aa121d43a37d2c2f4d786662ec0485bba381477b97d75e3R1) where test cases are not written
and i have reviewed the PR multiple times it is not about coverage but proper assertions to be added without assertions coverage is just number to show.

**Reply by santhosh-tarento** [reply]
Thanks, yeah we will check and comeback

**Reply by santhosh-tarento** [reply]
@HarishGangula  I have made the changes, please have look into it, https://github.com/deviigopinath/knowledge-platform-tools-fork/pull/1/commits

**Reply by thillaibirla** [reply]
@HarishGangula - has this been merged?

**Comment by jayesh9747** [comment]
@santhosh-tarento Could you please guide me how to search in Hindi in full-text-search in ElasticSearch?

**Reply by santhosh-tarento** [reply]
@jayesh9747 what were you expecting here. this thread is related to taxonomy editor

**Reply by jayesh9747** [reply]
sorry but , i want to ask assessment on  service thread
